### DISCUSSION TITLE
Equivalent of WPF GlyphTypeface (loaded from URI)

### DISCUSSION THREAD

**Discussion by gebodal** [AvaloniaUI/Avalonia#15697] [Q&A]
Hello all,

I am trying the classic, to port a WPF application to Avalonia. In this project, there is a need to draw and position individual font glyphs (by glyph index in the font file, rather than unicode codepoint), and I was doing this via the WPF `GlyphTypeface` class, using something like the following:

```
ushort glyphIndex = 10;
double fontSize = 12;

Uri fontUri = new Uri("file://path/to/my/font.ttf"); // In WPF, this also supports TrueType/OpenType font collections, using "#1" at the end, or similar

GlyphTypeface glyphs = new GlyphTypeface(fontUri);

Geometry geometry = glyphs .GetGlyphOutline(glyphIndex, fontsize, fontsize);

// This then gets added into a `GeometryDrawing` or similar to be displayed on a canvas
```

The fonts have to be loaded in dynamically, and cannot be made part of the application. And it needs to support font collection files (`.ttc`, `.otc`).

I found a reference to a GlyphTypeface class in Avalonia using Google, [here](https://reference.avaloniaui.net/api/Avalonia.Media/GlyphTypeface/), but it doesn't seem to be in my installed version of Avalonia (11.0.10) - but there is an `IGlyphTypeface`?

Similar questions have been [asked](https://github.com/AvaloniaUI/Avalonia/discussions/10302) [before](https://github.com/AvaloniaUI/Avalonia/discussions/7316), but I don't particularly understand the answers (or there was no satisfying answer), or what level of implementation has come after. There are some code snippets [here](https://github.com/AvaloniaUI/Avalonia/discussions/13563#discussioncomment-8957844), but I'm unsure where they're supposed to be inserted? (Into my own code, or should I be downloading and modifying Avalonia?)

Does anyone have any idea if this can be done in Avalonia currently? And if so, would you be able to share a code snippet or two?

Many thanks in advance for any help you can provide.

**Comment by LaurentInSeattle** [comment]
After looking at the docs, I would try the following: 

Create a Typeface from Family, Weight, Style
Create a GlyphTypeface from Typeface
Get the Glyphs from the GlyphTypeface
Create a GlyphRun from your selected glyphs
Create a GlyphRunDrawing object from the GlyphRun
Use it in a DrawingGroup or draw it directly if you have a DrawingContext 

HTH
Warning => Years of XP with WPF, but still a noob with Avalonia. I might be wrong...

**Reply by gebodal** [reply]
Hi, thanks for the response!

Do you have any idea how one creates a Typeface/FontFamily from a non-resource font file?

**Comment by Gillibald** [comment]
We can add https://learn.microsoft.com/en-us/dotnet/api/system.windows.media.glyphtypeface.getglyphoutline?view=windowsdesktop-8.0

We currently only support creating a geometry from FormattedText.

Loading typeface from file is currently only possible via some custom code. Direct support can be added.

**Reply by gebodal** [reply]
Hi, thanks for getting back on this!

One of the iterations of my current project used FormattedText for this part, so I can work with that, I think.

This custom code you mention, do you happen to have a code snippet? Would it be something I need to do inside the Avalonia source code? (Which I can do, but I'd rather not have to!)

**Reply by Gillibald** [reply]
https://github.com/AvaloniaUI/Avalonia/discussions/13563#discussioncomment-8957844

**Reply by gebodal** [reply]
Ah, yes, I saw that one, but it just says "Can modify", and I can't work out which parts of the code are being modified. Are `EmbeddedFontCollection` and `FontManager` existing classes that need altering or extending, or am I creating new classes within my project...? I got a bit lost here.
